### PR TITLE
Ajax Widget

### PR DESCRIPTION
**Pull request by qsomazzi**
The goal of this PR is to add a very simple widget that will fit most of needs.
This widget need an Url (can be external) which going to be called to populate dashboard data.
The render is fully customisable and the widget provide some defaults values
capture d ecran 2015-08-06 a 13 54 40
capture d ecran 2015-08-06 a 13 54 51

### CODE REVIEW

File: Block/AjaxBlockService.php [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+use Sonata\BlockBundle\Block\BlockContextInterface;
+use Sonata\BlockBundle\Model\BlockInterface;
+use Sonata\CoreBundle\Model\Metadata;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\OptionsResolver\OptionsResolverInterface;
+
+/**
+ * Class AjaxBlockService
+ *
+ * @package Sonata\DashboardBundle\Block
```
Comment (Soullivaneuh) [inline Block/AjaxBlockService.php]: No `@package`.

File: Block/AjaxBlockService.php [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+use Sonata\AdminBundle\Form\FormMapper;
+use Sonata\BlockBundle\Block\BaseBlockService;
+use Sonata\BlockBundle\Block\BlockContextInterface;
+use Sonata\BlockBundle\Model\BlockInterface;
+use Sonata\CoreBundle\Model\Metadata;
+use Symfony\Component\HttpFoundation\Response;
+use Symfony\Component\OptionsResolver\OptionsResolverInterface;
+
+/**
+ * Class AjaxBlockService
```
Comment (Soullivaneuh) [inline Block/AjaxBlockService.php]: Don't need a comment if it just for naming a class.

File: Block/AjaxBlockService.php [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+            'navy'       => 'Navy',
+            'teal'       => 'Teal',
+            'olive'      => 'Olive',
+            'lime'       => 'Lime',
+            'orange'     => 'Orange',
+            'fuchsia'    => 'Fuchsia',
+            'purple'     => 'Purple',
+            'maroon'     => 'Maroon',
+            'black'      => 'Black',
+            'gray'       => 'Gray',
```
Comment (Soullivaneuh) [inline Block/AjaxBlockService.php]: Is that colors of admin-lte theme?

If yes, it could be used elsewhere. Maybe a common access would be better, on core-bundle for example.

File: Resources/config/block.xml [hunk 4]
```diff
@@ -5,6 +5,7 @@
 
     <parameters>
         <parameter key="sonata.dashboard.block.container.class">Sonata\DashboardBundle\Block\ContainerBlockService</parameter>
+        <parameter key="sonata.dashboard.block.ajax.class">Sonata\DashboardBundle\Block\AjaxBlockService</parameter>
```
Comment (Soullivaneuh) [inline Resources/config/block.xml]: Using class parameter is not recommended by Symfony anymore.

File: Resources/views/Block/ajax.html.twig [hunk 5]
```diff
@@ -0,0 +1,27 @@
+{#
+This file is part of the "Outil Auteur" project.
```
Comment (Soullivaneuh) [inline Resources/views/Block/ajax.html.twig]: What is this header? ping @rande

Comment (qsomazzi) [inline Resources/views/Block/ajax.html.twig]: my bad .... a nasty copy/paste

File: Block/AjaxBlockService.php (hunk 3 above)
Comment (qsomazzi) [inline Block/AjaxBlockService.php]: Yes it's only adminLte colors.

At first I wanted to make a colorpicker like in "https://almsaeedstudio.com/themes/AdminLTE/pages/forms/advanced.html"
Bug i thought doing a v1 simpler

Comment (Soullivaneuh) [inline Block/AjaxBlockService.php]: Yeah. But I'm just talking about colors definition itself. Could be placed elsewhere, what do you think?

Comment (qsomazzi) [inline Block/AjaxBlockService.php]: Yes you're right, in admin bundle that will be better no ?

Comment (Soullivaneuh) [inline Block/AjaxBlockService.php]: I hesitate between admin or core.

@rande what do you think?

Comment (qsomazzi) [inline Block/AjaxBlockService.php]: @Soullivaneuh a trait like this one will do the trick ? 
https://github.com/qsomazzi/portal.v2/blob/master/src/Portal/Component/Core/Color.php
What do you think ?

Waiting for @rande opinion to see where we put it.

Comment (Soullivaneuh) [inline Block/AjaxBlockService.php]: I don't think `trait` is a best way to handle this problematic.

I would prefer a class with constant for example.

Comment (qsomazzi) [inline Block/AjaxBlockService.php]: @Soullivaneuh I've discuss about this with rande, you're right, I'm going to make a form_type 'simple_color_picker' inside core_bundle. I've never done that so I will check how to do this as soon as i can

Comment (qsomazzi) [inline Block/AjaxBlockService.php]: @Soullivaneuh https://github.com/sonata-project/SonataCoreBundle/pull/182

**Comment by qsomazzi** [comment]
@Soullivaneuh @rande up

File: Resources/views/Block/ajax.html.twig [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+            <span class="info-box-icon bg-{{ block.settings.color }}"><i class="{{ block.settings.icon }}"></i></span>
+            <div class="info-box-content">
+                <span class="info-box-text">{{ block.settings.title }}</span>
+                <span class="info-box-number"></span>
+            </div>
+        </div>
+    </div>
+    <script>
+        $.ajax({
+            url: "{{ block.settings.url }}"
```
Comment (OskarStark) [inline Resources/views/Block/ajax.html.twig]: no escaping needed? what if you have query parameters? does it work?

File: Tests/Block/AjaxBlockServiceTest.php [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+namespace Sonata\DashboardBundle\Tests\Block;
+
+use Sonata\BlockBundle\Block\BlockContext;
+use Sonata\BlockBundle\Model\Block;
+use Sonata\BlockBundle\Tests\Block\Service\FakeTemplating;
+use Sonata\DashboardBundle\Block\AjaxBlockService;
+
+/**
+ * Test Container Block service.
```
Comment (OskarStark) [inline Tests/Block/AjaxBlockServiceTest.php]: wrong description

**Comment by OskarStark** [comment]
I like this PR :+1:

**Comment by core23** [comment]
Any news on this @qsomazzi ?

Could we *use* this bundle with this PR?

**Comment by qsomazzi** [comment]
Hello @core23, I'm sorry i don't have time to work on this bundle for the moment, feel free to continue it as your wish. We can use this bundle as it is yes but some work is needed to make it fully operationnal.

File: Block/AjaxBlockService.php [hunk 8]
```diff
@@ -0,0 +1,86 @@
...
+            'class'    => 'col-md-3 col-sm-6 col-xs-12',
+        ));
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getName()
+    {
+        return 'Ajax (core)';
```
Comment (core23) [inline Block/AjaxBlockService.php]: Remove this line, so we can use the block id for translated names.

File: Block/AjaxBlockService.php [hunk 9]
```diff
@@ -0,0 +1,86 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function buildEditForm(FormMapper $form, BlockInterface $block)
+    {
+        $form->add('settings', 'sonata_type_immutable_array', array(
+            'label' => false,
+            'keys'  => array(
+                array('title', 'text', array()),
```
Comment (core23) [inline Block/AjaxBlockService.php]: Can you add ``label`` and ``required`` options for each key.

**Comment by Soullivaneuh** [comment]
According to the new [Sonata version management](https://github.com/sonata-project/SonataAdminBundle/issues/3053) and next major [release plan](https://github.com/sonata-project/SonataAdminBundle/issues/3731), this project has been refactored regarding branching and versioning.

If you see this message, your PR concerns a patch or a minor release and is not targeting the right branch.

So I'm closing this one, but don't see it as a refusal. If you think your work is still relevant and want to continue, feel free to reopen it on the right branch (e.g. the default one).

Regards.
